File: tools/vnext_contracts/stage0/decision_closure/validate.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Independent Ruby encoder and semantic validator for Stage-0 Decision closure.

require "digest"
require "json"

EXTERNAL_DOMAIN = "maestro.vnext.external-design-authority-closure.v1"
DECISION_DOMAIN = "maestro.vnext.decision-closure.v1"
SUCCESSOR_DECISION_STORE = "18f14bce862e15be09c9d88155d62627582df50c7754e2e8e1d6f6bee8f7d522"
SUCCESSOR_HEADS = {
  "dec-canonical-authority-materialization-df3b" => ["locked", "0d7c406f68f04fdf47ce00d56e8189b54159f164323c9511504790b941f715d0", "624f81c44b1a6459bc13472df05f547276d694e0f38c7216bb8df732aa3418cf"],
  "dec-canonical-execution-h3-verified-0939" => ["locked", "b5935c389182a7f3ec6447fb2a13dcb70e912108b399d0b1d25fee5f132186a7", "a98f1fdb95fcb3f2604936f50e9aa6661ad75bd51469d576e49239c5a6138307"],
  "dec-canonical-foundation-descriptor-a128" => ["locked", "17fb79ef9bc74cf3838d869bf5fb3b0ae0e9ae017670ca7cb207aeb8105c234e", "59fc4db26ec24f2f2ddc2df5cd70462f767e5d7e2d81644edc11a61c7fb7b26c"],
  "dec-canonical-installation-consumer-c1fe" => ["locked", "aaba56a8f34fb293a68f26743fbf4ef879d9f5a399a4eb45da74eed70a509e53", "5f35840fed183b406baab4cf9044ab05e3677f7061798c7949e80a868d2cd466"],
  "dec-canonical-non-action-protected-90a9" => ["locked", "8c6be56db78d8695b4e85e09fc4217257fee0b2dce0f5b5be8ef10230f24c20e", "7f0ea93dddef6354183b48cec27f6dee47f802688956bf552e3cb64ecca88f81"],
  "dec-canonical-trusted-host-protected-1fbc" => ["locked", "e572dc28e0c811c81207558e64b0372f757a873122b7f537f6354af819f118d8", "e6e84dea058097be48312ef98154958246763bac7f38d877ea04aee4af030d99"]
}.freeze

class Bytes
  attr_reader :value
  def initialize(hex)
    @value = [hex].pack("H*")
  end
end

def head(major, value)
  raise "unsigned u64 required" unless value.is_a?(Integer) && value.between?(0, 0xffffffffffffffff)
  return [(major << 5) | value].pack("C") if value < 24
  return [(major << 5) | 24, value].pack("CC") if value <= 0xff
  return [(major << 5) | 25, value].pack("Cn") if value <= 0xffff
  return [(major << 5) | 26, value].pack("CN") if value <= 0xffffffff

  [(major << 5) | 27, value].pack("CQ>")
end

def encode(value)
  case value
  when false then "\xf4".b
  when true then "\xf5".b
  when Integer then head(0, value)
  when Bytes then head(2, value.value.bytesize) + value.value
  when String
    raw = value.encode(Encoding::US_ASCII).b
    head(3, raw.bytesize) + raw
  when Array then head(4, value.length) + value.map { |item| encode(item) }.join
  else raise "unsupported deterministic CBOR value #{value.class}"
  end
end

def optional(value)
  value.nil? ? [0] : [1, value]
end

def raw_bytes(record)
  Bytes.new(record.fetch("raw_record_bytes").fetch("bytes"))
end

def external_record(record)
  [record.fetch("id"), record.fetch("terminal_status"), Bytes.new(record.fetch("raw_record_sha256")), Bytes.new(record.fetch("raw_body_sha256")), record.fetch("raw_supersedes"), record.fetch("raw_superseded_by"), record.fetch("external_authoring_disposition"), optional(record["normalized_successor"]), record.fetch("consequence_classification"), optional(record["rationale_disposition"]), record.fetch("materialization_ids").map { |item| Bytes.new(item) }, record.fetch("derived_effect_status"), raw_bytes(record)]
end

def decision_record(record)
  [record.fetch("id"), record.fetch("terminal_status"), Bytes.new(record.fetch("raw_record_sha256")), Bytes.new(record.fetch("raw_body_sha256")), record.fetch("raw_supersedes"), record.fetch("raw_superseded_by"), record.fetch("external_authoring_disposition"), optional(record["normalized_successor"]), record.fetch("consequence_classification"), optional(record["rationale_disposition"]), record.fetch("materialization_ids").map { |item| Bytes.new(item) }, record.fetch("derived_effect_status")]
end

def materialization(item)
  [Bytes.new(item.fetch("id")), item.fetch("artifact_id"), item.fetch("component_kind_tag"), 0, item.fetch("decision_sources").map { |source| [source.fetch("id"), Bytes.new(source.fetch("body_sha256"))] }]
end

def canonical_value(document, external)
  lineage = document.fetch("lineage")
  records = document.fetch("records").map { |item| external ? external_record(item) : decision_record(item) }
  ignored = lineage.fetch("ignored_unilateral_claims").map { |item| [item.fetch("source"), item.fetch("claimed_predecessor")] }
  composites = lineage.fetch("composite_external_heads").map { |item| [item.fetch("id"), item.fetch("raw_supersedes")] }
  value = [1, records, document.fetch("materializations").map { |item| materialization(item) }, ignored, composites]
  value << lineage.fetch("recognized_external_composite_heads") if external
  value
end

def validate(document, external)
  raise "closure not closed" unless document.fetch("closure_state") == "closed"
  decisions_sha256 = document.fetch("source_provenance_excluded_from_identity").fetch("decisions_sha256")
  summary = case decisions_sha256
  when "1f97e67b156d5a17d13b94ff955ad17efeb3bb71a4b74b1aec14e20dac1100dd"
    { "total" => 207, "locked" => 112, "superseded" => 95, "open" => 0, "material" => 204, "rationale_only" => 3, "unresolved_mappings" => 0, "pending_component_slots" => 109, "normalized_one_to_one_edges" => 23 }
  when SUCCESSOR_DECISION_STORE
    { "total" => 213, "locked" => 117, "superseded" => 96, "open" => 0, "material" => 210, "rationale_only" => 3, "unresolved_mappings" => 0, "pending_component_slots" => 114, "normalized_one_to_one_edges" => 24 }
  else
    raise "unknown Decision-store provenance"
  end
  raise "summary drift" unless document.fetch("summary") == summary
  materialization_base = { "kind" => "initial_external_design_closure", "decision_closure_id" => external ? document.fetch("decision_closure_reference") : document.fetch("identity") }
  root_assembly = { "state" => "pending_exact_component_resolution", "resolved_component_ids" => [], "materialization_base" => materialization_base, "candidate_root_after" => nil, "finalization_manifest_id" => nil }
  raise "fabricated or incomplete root resolution" unless document.fetch("root_assembly") == root_assembly
  records = document.fetch("records")
  ids = records.map { |item| item.fetch("id") }
  raise "records unsorted or duplicate" unless ids == ids.sort && ids.uniq.length == summary.fetch("total")
  index = records.to_h { |record| [record.fetch("id"), record] }
  if decisions_sha256 == SUCCESSOR_DECISION_STORE
    manifest_bytes = records.map do |record|
      [
        record.fetch("id"),
        record.fetch("terminal_status"),
        record.fetch("raw_record_sha256"),
        record.fetch("raw_body_sha256")
      ].join("\t")
    end.join("\n") + "\n"
    raise "all-Decision successor manifest reconstruction mismatch" unless
      Digest::SHA256.hexdigest(manifest_bytes) == SUCCESSOR_DECISION_STORE
    SUCCESSOR_HEADS.each do |decision_id, expected|
      record = index.fetch(decision_id) { raise "missing successor Decision head: #{decision_id}" }
      actual = [record.fetch("terminal_status"), record.fetch("raw_record_sha256"), record.fetch("raw_body_sha256")]
      raise "substituted successor Decision head: #{decision_id}" unless actual == expected
    end
    ignored = document.fetch("lineage").fetch("ignored_unilateral_claims").map do |item|
      [item.fetch("source"), item.fetch("claimed_predecessor")]
    end
    unless ignored.include?(["dec-canonical-non-action-protected-90a9", "dec-canonical-trusted-host-protected-1fbc"])
      raise "missing protected-diagnostic unilateral-claim refusal"
    end
  end
  records.each do |record|
    seen = []
    current = record.fetch("id")
    until current.nil?
      raise "normalized successor cycle" if seen.include?(current)

      seen << current
      current = index.fetch(current)["normalized_successor"]
    end
  end
  materialization_rows = document.fetch("materializations")
  raise "duplicate materialization" unless materialization_rows.map { |item| item.fetch("id") }.uniq.length == materialization_rows.length
  used = []
  records.each do |record|
    raise "raw record mismatch" if external && Digest::SHA256.hexdigest(raw_bytes(record).value) != record.fetch("raw_record_sha256")
    raise "invalid terminal status" unless %w[locked superseded].include?(record.fetch("terminal_status"))
    raise "raw lineage omission" if record.fetch("terminal_status") == "superseded" && record.fetch("raw_superseded_by").empty?
    raise "composite promotion" if record.fetch("external_authoring_disposition") == "composite_external_authoring" && !record["normalized_successor"].nil?
    raise "unilateral repair" if record.fetch("external_authoring_disposition") == "unilateral_raw_claim" && !record["normalized_successor"].nil?
    if record.fetch("consequence_classification") == "rationale_only"
      raise "missing rationale disposition" if record["rationale_disposition"].nil? || !record.fetch("materialization_ids").empty?
    else
      raise "missing materialization" if record.fetch("materialization_ids").empty?
      raise "effect-live coverage missing" unless %w[unapplied superseded_but_effect_live].include?(record.fetch("derived_effect_status"))
    end
    successor = record["normalized_successor"]
    raise "unknown successor" if !successor.nil? && !index.key?(successor)
  end
  materialization_rows.each do |item|
    raise "materialization base drift" unless item.fetch("materialization_base") == materialization_base
    raise "materialization must remain pending exact root resolution" unless item.fetch("binding_state") == "required_component_slot_pending" && !item.key?("before_root_id") && %w[exact_component_id after_root_id finalization_manifest_id].all? { |key| item[key].nil? }
    sources = item.fetch("decision_sources")
    source_ids = sources.map { |source| source.fetch("id") }
    raise "duplicate materialization source" unless source_ids == source_ids.sort && source_ids.uniq.length == source_ids.length
    sources.each do |source|
      record = index.fetch(source.fetch("id"))
      raise "stale materialization" unless source.fetch("body_sha256") == record.fetch("raw_body_sha256")
      raise "nonreciprocal materialization" unless record.fetch("materialization_ids").include?(item.fetch("id"))
      used << source.fetch("id")
    end
  end
  expected = records.select { |record| record.fetch("consequence_classification") == "material" }.map { |record| record.fetch("id") }
  raise "incomplete materialization closure" unless used.sort == expected.sort
  value = canonical_value(document, external)
  encoded = encode(value)
  domain = external ? EXTERNAL_DOMAIN : DECISION_DOMAIN
  identity = Digest::SHA256.hexdigest(encode([domain, value]))
  raise "identity mismatch" unless document.fetch("identity") == "sha256:#{identity}"
  raise "CBOR hash mismatch" unless document.fetch("canonical_cbor_sha256") == Digest::SHA256.hexdigest(encoded)
  [identity, encoded]
end

root = ENV.fetch("STAGE0_DECISION_CLOSURE_ROOT", File.expand_path("../../../../contracts/vnext/stage0/decision-closure", __dir__))
external = JSON.parse(File.read(File.join(root, "external-design-authority-closure.v1.json"), encoding: Encoding::US_ASCII))
decision = JSON.parse(File.read(File.join(root, "decision-closure.v1.json"), encoding: Encoding::US_ASCII))
external_id, external_cbor = validate(external, true)
decision_id, decision_cbor = validate(decision, false)
raise "external CBOR file mismatch" unless external_cbor == File.binread(File.join(root, "external-design-authority-closure.v1.cbor"))
raise "Decision CBOR file mismatch" unless decision_cbor == File.binread(File.join(root, "decision-closure.v1.cbor"))
puts JSON.generate({ "external_closure_id" => "sha256:#{external_id}", "decision_closure_id" => "sha256:#{decision_id}", "encoder" => "ruby", "semantic_validation" => "pass" })

File: tools/vnext_contracts/stage0/decision_closure/validate.py
#!/usr/bin/env python3
"""Independent Stage-0 Decision closure CBOR encoder and semantic validator."""

from __future__ import annotations

import hashlib
import json
import os
import sys
from pathlib import Path


EXTERNAL_DOMAIN = "maestro.vnext.external-design-authority-closure.v1"
DECISION_DOMAIN = "maestro.vnext.decision-closure.v1"
SUCCESSOR_DECISION_STORE = "18f14bce862e15be09c9d88155d62627582df50c7754e2e8e1d6f6bee8f7d522"
SUCCESSOR_HEADS = {
    "dec-canonical-authority-materialization-df3b": ("locked", "0d7c406f68f04fdf47ce00d56e8189b54159f164323c9511504790b941f715d0", "624f81c44b1a6459bc13472df05f547276d694e0f38c7216bb8df732aa3418cf"),
    "dec-canonical-execution-h3-verified-0939": ("locked", "b5935c389182a7f3ec6447fb2a13dcb70e912108b399d0b1d25fee5f132186a7", "a98f1fdb95fcb3f2604936f50e9aa6661ad75bd51469d576e49239c5a6138307"),
    "dec-canonical-foundation-descriptor-a128": ("locked", "17fb79ef9bc74cf3838d869bf5fb3b0ae0e9ae017670ca7cb207aeb8105c234e", "59fc4db26ec24f2f2ddc2df5cd70462f767e5d7e2d81644edc11a61c7fb7b26c"),
    "dec-canonical-installation-consumer-c1fe": ("locked", "aaba56a8f34fb293a68f26743fbf4ef879d9f5a399a4eb45da74eed70a509e53", "5f35840fed183b406baab4cf9044ab05e3677f7061798c7949e80a868d2cd466"),
    "dec-canonical-non-action-protected-90a9": ("locked", "8c6be56db78d8695b4e85e09fc4217257fee0b2dce0f5b5be8ef10230f24c20e", "7f0ea93dddef6354183b48cec27f6dee47f802688956bf552e3cb64ecca88f81"),
    "dec-canonical-trusted-host-protected-1fbc": ("locked", "e572dc28e0c811c81207558e64b0372f757a873122b7f537f6354af819f118d8", "e6e84dea058097be48312ef98154958246763bac7f38d877ea04aee4af030d99"),
}


def head(major: int, value: int) -> bytes:
    if not 0 <= value <= 0xFFFFFFFFFFFFFFFF:
        raise ValueError("unsigned u64 required")
    if value < 24:
        return bytes([(major << 5) | value])
    if value <= 0xFF:
        return bytes([(major << 5) | 24, value])
    if value <= 0xFFFF:
        return bytes([(major << 5) | 25]) + value.to_bytes(2, "big")
    if value <= 0xFFFFFFFF:
        return bytes([(major << 5) | 26]) + value.to_bytes(4, "big")
    return bytes([(major << 5) | 27]) + value.to_bytes(8, "big")


def encode(value: object) -> bytes:
    if value is False:
        return b"\xf4"
    if value is True:
        return b"\xf5"
    if isinstance(value, int) and not isinstance(value, bool):
        return head(0, value)
    if isinstance(value, str):
        raw = value.encode("ascii")
        return head(3, len(raw)) + raw
    if isinstance(value, list):
        return head(4, len(value)) + b"".join(encode(item) for item in value)
    if isinstance(value, dict) and list(value) == ["bytes"]:
        raw = bytes.fromhex(value["bytes"])
        return head(2, len(raw)) + raw
    raise ValueError(f"value outside deterministic CBOR subset: {value!r}")


def raw_bytes(record: dict[str, object]) -> dict[str, str]:
    return {"bytes": record["raw_record_bytes"]["bytes"]}  # type: ignore[index]


def optional(value: object) -> list[object]:
    return [0] if value is None else [1, value]


def external_record(record: dict[str, object]) -> list[object]:
    return [
        record["id"], record["terminal_status"], {"bytes": record["raw_record_sha256"]},
        {"bytes": record["raw_body_sha256"]}, record["raw_supersedes"],
        record["raw_superseded_by"], record["external_authoring_disposition"],
        optional(record["normalized_successor"]), record["consequence_classification"],
        optional(record["rationale_disposition"]),
        [{"bytes": value} for value in record["materialization_ids"]],
        record["derived_effect_status"], raw_bytes(record),
    ]


def decision_record(record: dict[str, object]) -> list[object]:
    return [
        record["id"], record["terminal_status"], {"bytes": record["raw_record_sha256"]},
        {"bytes": record["raw_body_sha256"]}, record["raw_supersedes"],
        record["raw_superseded_by"], record["external_authoring_disposition"],
        optional(record["normalized_successor"]), record["consequence_classification"],
        optional(record["rationale_disposition"]),
        [{"bytes": value} for value in record["materialization_ids"]],
        record["derived_effect_status"],
    ]


def materialization(item: dict[str, object]) -> list[object]:
    return [
        {"bytes": item["id"]}, item["artifact_id"], item["component_kind_tag"], 0,
        [[source["id"], {"bytes": source["body_sha256"]}] for source in item["decision_sources"]],
    ]


def value(document: dict[str, object], external: bool) -> list[object]:
    lineage = document["lineage"]
    records = [external_record(item) if external else decision_record(item) for item in document["records"]]
    materials = [materialization(item) for item in document["materializations"]]
    ignored = [[item["source"], item["claimed_predecessor"]] for item in lineage["ignored_unilateral_claims"]]
    composites = [[item["id"], item["raw_supersedes"]] for item in lineage["composite_external_heads"]]
    base: list[object] = [1, records, materials, ignored, composites]
    if external:
        base.append(lineage["recognized_external_composite_heads"])
    return base


def validate(document: dict[str, object], external: bool) -> tuple[str, bytes]:
    summary = document["summary"]
    records = document["records"]
    provenance = document["source_provenance_excluded_from_identity"]
    decisions_sha256 = provenance["decisions_sha256"]
    if decisions_sha256 == "1f97e67b156d5a17d13b94ff955ad17efeb3bb71a4b74b1aec14e20dac1100dd":
        expected = {"total": 207, "locked": 112, "superseded": 95, "open": 0, "material": 204, "rationale_only": 3, "unresolved_mappings": 0, "pending_component_slots": 109, "normalized_one_to_one_edges": 23}
    elif decisions_sha256 == SUCCESSOR_DECISION_STORE:
        expected = {"total": 213, "locked": 117, "superseded": 96, "open": 0, "material": 210, "rationale_only": 3, "unresolved_mappings": 0, "pending_component_slots": 114, "normalized_one_to_one_edges": 24}
    else:
        raise ValueError("unknown Decision-store provenance")
    if document["closure_state"] != "closed":
        raise ValueError("closure is not closed")
    if [item["id"] for item in records] != sorted(item["id"] for item in records):
        raise ValueError("records are not sorted")
    if len({item["id"] for item in records}) != expected["total"]:
        raise ValueError("omitted or duplicate Decision record")
    if summary != expected:
        raise ValueError("summary drift")
    if decisions_sha256 == SUCCESSOR_DECISION_STORE:
        manifest_bytes = (
            "".join(
                f"{item['id']}\t{item['terminal_status']}\t"
                f"{item['raw_record_sha256']}\t{item['raw_body_sha256']}\n"
                for item in records
            )
        ).encode("ascii")
        if hashlib.sha256(manifest_bytes).hexdigest() != SUCCESSOR_DECISION_STORE:
            raise ValueError("all-Decision successor manifest reconstruction mismatch")
        index = {item["id"]: item for item in records}
        for decision_id, (status, record_sha, body_sha) in SUCCESSOR_HEADS.items():
            item = index.get(decision_id)
            if item is None:
                raise ValueError(f"missing successor Decision head: {decision_id}")
            if (
                item["terminal_status"] != status
                or item["raw_record_sha256"] != record_sha
                or item["raw_body_sha256"] != body_sha
            ):
                raise ValueError(f"substituted successor Decision head: {decision_id}")
        ignored = {
            (item["source"], item["claimed_predecessor"])
            for item in document["lineage"]["ignored_unilateral_claims"]
        }
        if (
            "dec-canonical-trusted-host-protected-1fbc",
            "dec-canonical-non-action-protected-90a9",
        ) not in ignored:
            raise ValueError("missing protected-diagnostic unilateral-claim refusal")
    materialization_base = {
        "kind": "initial_external_design_closure",
        "decision_closure_id": document["decision_closure_reference"] if external else document["identity"],
    }
    root_assembly = document["root_assembly"]
    if root_assembly != {"state": "pending_exact_component_resolution", "resolved_component_ids": [], "materialization_base": materialization_base, "candidate_root_after": None, "finalization_manifest_id": None}:
        raise ValueError("fabricated or incomplete root resolution")
    ids = {item["id"] for item in records}
    successors = {item["id"]: item["normalized_successor"] for item in records}
    for start in successors:
        seen: set[str] = set()
        current: str | None = start
        while current is not None:
            if current in seen:
                raise ValueError("normalized successor cycle")
            seen.add(current)
            current = successors[current]
    materialization_rows = document["materializations"]
    if len({item["id"] for item in materialization_rows}) != len(materialization_rows):
        raise ValueError("duplicate materialization")
    material_ids: dict[str, dict[str, object]] = {item["id"]: item for item in materialization_rows}
    seen_sources: set[str] = set()
    for item in records:
        if external:
            raw = bytes.fromhex(item["raw_record_bytes"]["bytes"])
            if hashlib.sha256(raw).hexdigest() != item["raw_record_sha256"]:
                raise ValueError("raw record hash changed")
        if item["terminal_status"] not in {"locked", "superseded"}:
            raise ValueError("open or invalid terminal status")
        if item["terminal_status"] == "superseded" and not item["raw_superseded_by"]:
            raise ValueError("raw lineage omission")
        if item["external_authoring_disposition"] == "composite_external_authoring" and item["normalized_successor"] is not None:
            raise ValueError("composite promotion")
        if item["external_authoring_disposition"] == "unilateral_raw_claim" and item["normalized_successor"] is not None:
            raise ValueError("unilateral repair")
        if item["consequence_classification"] == "rationale_only":
            if item["rationale_disposition"] is None or item["materialization_ids"]:
                raise ValueError("missing rationale disposition")
        else:
            if not item["materialization_ids"]:
                raise ValueError("missing materialization")
            if item["derived_effect_status"] not in {"unapplied", "superseded_but_effect_live"}:
                raise ValueError("missing effect-live coverage")
        successor = item["normalized_successor"]
        if successor is not None and successor not in ids:
            raise ValueError("unknown normalized successor")
    for material_id, item in material_ids.items():
        if item["materialization_base"] != materialization_base:
            raise ValueError("materialization base drift")
        if item["binding_state"] != "required_component_slot_pending" or "before_root_id" in item or any(item[key] is not None for key in ("exact_component_id", "after_root_id", "finalization_manifest_id")):
            raise ValueError("materialization must remain pending exact root resolution")
        source_ids = [source["id"] for source in item["decision_sources"]]
        if source_ids != sorted(source_ids) or len(source_ids) != len(set(source_ids)):
            raise ValueError("duplicate or unordered materialization source")
        for source in item["decision_sources"]:
            record = next(row for row in records if row["id"] == source["id"])
            if source["body_sha256"] != record["raw_body_sha256"]:
                raise ValueError("stale materialization")
            if material_id not in record["materialization_ids"]:
                raise ValueError("non-reciprocal materialization")
            seen_sources.add(source["id"])
    if seen_sources != {item["id"] for item in records if item["consequence_classification"] == "material"}:
        raise ValueError("incomplete materialization closure")
    domain = EXTERNAL_DOMAIN if external else DECISION_DOMAIN
    encoded = encode(value(document, external))
    identity = hashlib.sha256(encode([domain, value(document, external)])).hexdigest()
    if document["identity"] != f"sha256:{identity}":
        raise ValueError("identity mismatch")
    if document["canonical_cbor_sha256"] != hashlib.sha256(encoded).hexdigest():
        raise ValueError("canonical CBOR mismatch")
    return identity, encoded


def main() -> int:
    root = Path(os.environ.get("STAGE0_DECISION_CLOSURE_ROOT", Path(__file__).resolve().parents[4] / "contracts/vnext/stage0/decision-closure"))
    external = json.loads((root / "external-design-authority-closure.v1.json").read_text(encoding="ascii"))
    decision = json.loads((root / "decision-closure.v1.json").read_text(encoding="ascii"))
    external_id, external_cbor = validate(external, True)
    decision_id, decision_cbor = validate(decision, False)
    if external_cbor != (root / "external-design-authority-closure.v1.cbor").read_bytes():
        raise ValueError("external CBOR file mismatch")
    if decision_cbor != (root / "decision-closure.v1.cbor").read_bytes():
        raise ValueError("Decision CBOR file mismatch")
    print(json.dumps({"external_closure_id": f"sha256:{external_id}", "decision_closure_id": f"sha256:{decision_id}", "encoder": "python", "semantic_validation": "pass"}, separators=(",", ":")))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
